include/LinearSolver.h

```diff
@@ -90,20 +91,22 @@ class LocalGraphArrays {
     }
   }
 
-  static size_t compute_row_pointers(Teuchos::ArrayRCP<size_t>& rowPtrs,
-                                   const Kokkos::View<size_t*,sierra::nalu::DeviceSpace>& rowLengths)
+  static size_t compute_row_pointers(Kokkos::View<size_t*>& rowPtrs,
```

Comment:
The device type is missing; did you mean to do that?

Reply:
I should probably specify device type for all views... In this case, rowPointers and colIndices are for passing into CrsGraph::setAllIndices, which takes these two types:
   const typename local_graph_type::row_map_type & rowPointers
   const typename local_graph_type::entries_type::non_const_type & columnIndices
I've had trouble finding exactly what those types need to be. Do they expect a particular device type? Can setAllIndices (and other CrsGraph methods) be called on either host or device (in gpu context)?
For now, our linear-system assembly is tied to the host, but in the future it may not be.

Reply:
@alanw0 wrote: 
> I've had trouble finding exactly what those types need to be.

Those are Kokkos::View specializations.  `local_graph_type` is a Kokkos::StaticCrsGraph specialization.  (Tpetra::CrsGraph declares the `local_graph_type` typedef.)

> Do they expect a particular device type?

Their device type comes from the Tpetra class' `device_type` typedef.

> Can setAllIndices (and other CrsGraph methods) be called on either host or device (in gpu context)?

  1. Tpetra class functions are currently all host only.
  2. `setAllIndices` is a coarse-grained operation and therefore to be called on host.

Reply:
@alanw0 wrote:
> For now, our linear-system assembly is tied to the host, but in the future it may not be.

If you intend to fill on host, you could use host Views here or even host mirror views.  Host mirror views would be `CudaUVMSpace` currently, so please be sure to fence after filling on host.

Reply:
@mhoemmen thanks for the info. given that Tpetra class functions are host only, I will use HostSpace for all of our views. Is there a plan for Tpetra functions to be device functions in the future? The stk mesh project has been experimenting with a 'ngp::Mesh' class which makes a device snapshot of the mesh, and has device functions for doing common mesh traversal/access operations. That's probably what nalu will use if/when we attempt to fill on device.

Reply:
@alanw0 wrote:
> given that Tpetra class functions are host only, I will use HostSpace for all of our views.

`setAllIndices` is a host function, yes, but it takes device Views.

Reply:
@alanw0 wrote:
> Is there a plan for Tpetra functions to be device functions in the future?

KokkosSparse::CrsMatrix has much of the functionality you need for sparse matrix assembly (with fixed graph structure -- sumInto/replace, not insert).  See e.g., `sumIntoValues` and `replaceValues`.

We have some plans to write easier-to-use graph construction interfaces, but for now, please see the Kokkos FENL example for thread-parallel graph construction.